packages/ws/src/ws/WebSocketShard.ts

```diff
@@ -0,0 +1,515 @@
+				inflate(decompressable, { chunkSize: 65535 }, (err, result) => {
+					if (err) {
+						return reject(err);
+					}
+
+					resolve(JSON.parse(this.textDecoder.decode(result)) as GatewayReceivePayload);
+				});
+			});
+		}
+
+		// Deal with gw wide zlib-stream compression
+		if (this.inflate) {
+			const l = decompressable.length;
+			const flush =
+				l >= 4 &&
+				decompressable[l - 4] === 0x00 &&
+				decompressable[l - 3] === 0x00 &&
+				decompressable[l - 2] === 0xff &&
+				decompressable[l - 1] === 0xff;
+
+			const zlib = await getZlibSync();
+			this.inflate.push(Buffer.from(decompressable), flush ? zlib!.Z_SYNC_FLUSH : zlib!.Z_NO_FLUSH);
+
+			if (this.inflate.err) {
+				this.emit('error', `${this.inflate.err}${this.inflate.msg ? `: ${this.inflate.msg}` : ''}`);
+			}
+
+			if (!flush) {
+				return null;
+			}
+
+			const { result } = this.inflate;
+			if (!result) {
+				return null;
+			}
+
+			return JSON.parse(typeof result === 'string' ? result : this.textDecoder.decode(result)) as GatewayReceivePayload;
+		}
+
+		this.debug([
+			'Received a message we were unable to decompress',
+			`isBinary: ${isBinary.toString()}`,
+			`useIdentifyCompress: ${this.useIdentifyCompress.toString()}`,
+			`inflate: ${Boolean(this.inflate).toString()}`,
+		]);
+
+		return null;
+	}
+
+	private async onMessage(data: RawData, isBinary: boolean) {
+		const payload = await this.unpackMessage(data as Buffer | ArrayBuffer, isBinary);
+		if (!payload) {
+			return;
+		}
+
+		switch (payload.op) {
+			case GatewayOpcodes.Dispatch: {
+				if (this.status === WebSocketShardStatus.Resuming) {
+					this.replayedEvents++;
+				}
+
+				switch (payload.t) {
+					case GatewayDispatchEvents.Ready: {
+						this.emit(WebSocketShardEvents.Ready);
+
+						this.session ??= {
+							sequence: payload.s,
+							sessionId: payload.d.session_id,
+							shardId: this.id,
+							shardCount: this.strategy.options.shardCount,
+						};
+
+						await this.strategy.updateSessionInfo(this.id, this.session);
+						break;
+					}
+
+					case GatewayDispatchEvents.Resumed: {
+						this.status = WebSocketShardStatus.Ready;
+						this.debug([`Resumed and replayed ${this.replayedEvents} events`]);
+						this.emit(WebSocketShardEvents.Resumed);
+						break;
+					}
+
+					default: {
+						break;
+					}
+				}
+
+				if (this.session) {
+					if (payload.s > this.session.sequence) {
+						this.session.sequence = payload.s;
+						await this.strategy.updateSessionInfo(this.id, this.session);
+					}
+				}
+
+				break;
+			}
+
+			case GatewayOpcodes.Heartbeat: {
+				await this.heartbeat(true);
+				break;
+			}
+
+			case GatewayOpcodes.Reconnect: {
+				await this.destroy({
+					reason: 'Told to reconnect by Discord',
+					recover: WebSocketShardDestroyRecovery.Reconnect,
+				});
+				break;
+			}
+
+			case GatewayOpcodes.InvalidSession: {
+				this.debug([`Invalid session; will attempt to resume: ${payload.d.toString()}`]);
```

Comment:
If resumable is true we can just send a resume payload directly, no need to close and reconnect

Reply:
oh? that's a TIL.